File: packages/firebase_remote_config/firebase_remote_config/ios/Classes/FirebaseRemoteConfigPlugin.m
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import <firebase_core/FLTFirebasePluginRegistry.h>
#import <Firebase/Firebase.h>

#import "FirebaseRemoteConfigPlugin.h"

NSString *const kFirebaseRemoteConfigChannelName = @"plugins.flutter.io/firebase_remote_config";

@interface FirebaseRemoteConfigPlugin ()
@property(nonatomic, retain) FlutterMethodChannel *channel;
@end

@implementation FirebaseRemoteConfigPlugin

+ (instancetype)sharedInstance {
  static dispatch_once_t onceToken;
  static FirebaseRemoteConfigPlugin *instance;

  dispatch_once(&onceToken, ^{
    instance = [[FirebaseRemoteConfigPlugin alloc] init];
    [[FLTFirebasePluginRegistry sharedInstance] registerFirebasePlugin:instance];
  });

  return instance;
}

- (instancetype)init {
  self = [super init];
  return self;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *channel =
      [FlutterMethodChannel methodChannelWithName:kFirebaseRemoteConfigChannelName
                                  binaryMessenger:[registrar messenger]];
  FirebaseRemoteConfigPlugin *instance = [FirebaseRemoteConfigPlugin sharedInstance];

  [registrar addMethodCallDelegate:instance channel:channel];

  SEL sel = NSSelectorFromString(@"registerLibrary:withVersion:");
  if ([FIRApp respondsToSelector:sel]) {
    [FIRApp performSelector:sel withObject:LIBRARY_NAME withObject:LIBRARY_VERSION];
  }
}

- (void)detachFromEngineForRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar {
  self.channel = nil;
}

- (FIRRemoteConfig *_Nullable)getFIRRemoteConfigFromArguments:(NSDictionary *)arguments {
  NSString *appName = arguments[@"appName"];
  FIRApp *app = [FLTFirebasePlugin firebaseAppNamed:appName];
  return [FIRRemoteConfig remoteConfigWithApp:app];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
  FIRRemoteConfig *remoteConfig = [self getFIRRemoteConfigFromArguments:call.arguments];
  if ([@"RemoteConfig#ensureInitialized" isEqualToString:call.method]) {
    [remoteConfig ensureInitializedWithCompletionHandler:^(NSError * _Nullable initializationError) {
      result(nil);
    }];
  } else if([@"RemoteConfig#activate" isEqualToString:call.method]) {
    [remoteConfig activateWithCompletion:^(BOOL configActivated, NSError *_Nullable activateError) {
      result(@(configActivated));
    }];
  } else if ([@"RemoteConfig#getAll" isEqualToString:call.method]) {
    NSDictionary *parameters = [self getAllParametersForInstance:remoteConfig];
    result(parameters);
  } else if ([@"RemoteConfig#fetch" isEqualToString:call.method] ) {
    [remoteConfig fetchWithCompletionHandler:^(FIRRemoteConfigFetchStatus status, NSError *error) {
      result(nil);
    }];
  } else if ([@"RemoteConfig#fetchAndActivate" isEqualToString:call.method]) {
    [remoteConfig fetchAndActivateWithCompletionHandler:^(FIRRemoteConfigFetchAndActivateStatus status, NSError *error) {
      if (status == FIRRemoteConfigFetchAndActivateStatusSuccessFetchedFromRemote) {
        result([NSNumber numberWithBool:TRUE]);
      } else {
        result([NSNumber numberWithBool:FALSE]);
      }
    }];
  } else if ([@"RemoteConfig#setConfigSettings" isEqualToString:call.method]) {
    NSNumber *fetchTimeout = call.arguments[@"fetchTimeout"];
    NSNumber *minimumFetchInterval = call.arguments[@"minimumFetchInterval"];
    FIRRemoteConfigSettings *remoteConfigSettings = [[FIRRemoteConfigSettings alloc] init];
    remoteConfigSettings.fetchTimeout = [fetchTimeout intValue];
    remoteConfigSettings.minimumFetchInterval = [minimumFetchInterval intValue];
    [remoteConfig setConfigSettings:remoteConfigSettings];
    result(nil);
  } else if ([@"RemoteConfig#setDefaults" isEqualToString:call.method]) {
    NSDictionary *defaults = call.arguments[@"defaults"];
    [remoteConfig setDefaults:defaults];
    result(nil);
  } else if ([@"RemoteConfig#getProperties" isEqualToString:call.method]) {
    NSDictionary *configProperties = [self configPropertiesForInstance:remoteConfig];
    result(configProperties);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (NSDictionary *)getAllParametersForInstance:(FIRRemoteConfig *)remoteConfig {
  NSMutableSet *keySet = [[NSMutableSet alloc] init];
  [keySet addObjectsFromArray:[remoteConfig allKeysFromSource:FIRRemoteConfigSourceStatic]];
  [keySet addObjectsFromArray:[remoteConfig allKeysFromSource:FIRRemoteConfigSourceDefault]];
  [keySet addObjectsFromArray:[remoteConfig allKeysFromSource:FIRRemoteConfigSourceRemote]];

  NSMutableDictionary *parameters = [[NSMutableDictionary alloc] init];
  for (NSString *key in keySet) {
    parameters[key] = [self createRemoteConfigValueDict: [remoteConfig configValueForKey:key]];
  }
  return parameters;
}

- (NSMutableDictionary *)createRemoteConfigValueDict:(FIRRemoteConfigValue *)remoteConfigValue {
  NSMutableDictionary *valueDict = [[NSMutableDictionary alloc] init];
  valueDict[@"value"] = [FlutterStandardTypedData typedDataWithBytes:[remoteConfigValue dataValue]];
  valueDict[@"source"] = [self mapValueSource:[remoteConfigValue source]];
  return valueDict;
}

- (NSString *)mapLastFetchStatus:(FIRRemoteConfigFetchStatus)status {
  if (status == FIRRemoteConfigFetchStatusSuccess) {
    return @"success";
  } else if (status == FIRRemoteConfigFetchStatusFailure) {
    return @"failure";
  } else if (status == FIRRemoteConfigFetchStatusThrottled) {
    return @"throttled";
  } else if (status == FIRRemoteConfigFetchStatusNoFetchYet) {
    return @"noFetchYet";
  } else {
    return @"failure";
  }
}

- (NSString *)mapValueSource:(FIRRemoteConfigSource)source {
  if (source == FIRRemoteConfigSourceStatic) {
    return @"static";
  } else if (source == FIRRemoteConfigSourceDefault) {
    return @"default";
  } else if (source == FIRRemoteConfigSourceRemote) {
    return @"remote";
  } else {
    return @"static";
  }
}

#pragma mark - FLTFirebasePlugin

- (void)didReinitializeFirebaseCore:(void (^)(void))completion {
  completion();
}

- (NSDictionary *_Nonnull)pluginConstantsForFIRApp:(FIRApp *)firebase_app {
  FIRRemoteConfig  *firebaseRemoteConfig = [FIRRemoteConfig remoteConfigWithApp:firebase_app];
  NSDictionary *configProperties = [self configPropertiesForInstance:firebaseRemoteConfig];

  NSMutableDictionary *configValues = [[NSMutableDictionary alloc] init];
  [configValues addEntriesFromDictionary:configProperties];
  [configValues setValue:[self getAllParametersForInstance:firebaseRemoteConfig] forKey:@"parameters"];
  return configValues;
}

- (NSDictionary *_Nonnull)configPropertiesForInstance:(FIRRemoteConfig *)remoteConfig {
  NSNumber *fetchTimeout = @([[remoteConfig configSettings] fetchTimeout]);
  NSNumber *minimumFetchInterval = @([[remoteConfig configSettings] minimumFetchInterval]);
  int lastFetchMillis = (int) ([[remoteConfig lastFetchTime] timeIntervalSince1970] * 1000);

  NSMutableDictionary *configProperties = [[NSMutableDictionary alloc] init];
  [configProperties setValue:@([fetchTimeout intValue]) forKey:@"fetchTimeout"];
  [configProperties setValue:@([minimumFetchInterval intValue]) forKey:@"minimumFetchInterval"];
  [configProperties setValue:@(lastFetchMillis) forKey:@"lastFetchTime"];
  [configProperties setValue:[self mapLastFetchStatus:[remoteConfig lastFetchStatus]] forKey:@"lastFetchStatus"];
  return configProperties;
}

- (NSString *_Nonnull)firebaseLibraryName {
  return LIBRARY_NAME;
}

- (NSString *_Nonnull)firebaseLibraryVersion {
  return LIBRARY_VERSION;
}

- (NSString *_Nonnull)flutterChannelName {
  return kFirebaseRemoteConfigChannelName;
}

@end

File: packages/firebase_remote_config/firebase_remote_config/ios/firebase_remote_config/Sources/firebase_remote_config/FirebaseRemoteConfigPlugin.swift
// Copyright 2025 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#if canImport(FlutterMacOS)
  import FlutterMacOS
#else
  import Flutter
#endif

#if canImport(firebase_core)
  import firebase_core
#else
  import firebase_core_shared
#endif
import FirebaseRemoteConfig

let kFirebaseRemoteConfigChannelName = "plugins.flutter.io/firebase_remote_config"
let kFirebaseRemoteConfigUpdatedChannelName = "plugins.flutter.io/firebase_remote_config_updated"

extension FlutterError: Error {}

public class FirebaseRemoteConfigPlugin: NSObject, FlutterPlugin, FlutterStreamHandler,
  FLTFirebasePluginProtocol, FirebaseRemoteConfigHostApi {
  private var listenersMap: [String: ConfigUpdateListenerRegistration] = [:]
  private var fetchAndActivateRetry = false

  static let shared: FirebaseRemoteConfigPlugin = {
    let instance = FirebaseRemoteConfigPlugin()
    FLTFirebasePluginRegistry.sharedInstance().register(instance)
    instance.fetchAndActivateRetry = false
    return instance
  }()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let binaryMessenger: FlutterBinaryMessenger

    #if os(macOS)
      binaryMessenger = registrar.messenger
    #elseif os(iOS)
      binaryMessenger = registrar.messenger()
    #endif

    let instance = shared
    FirebaseRemoteConfigHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: instance)

    let eventChannel = FlutterEventChannel(
      name: kFirebaseRemoteConfigUpdatedChannelName,
      binaryMessenger: binaryMessenger
    )
    eventChannel.setStreamHandler(instance)

    if FirebaseApp.responds(to: NSSelectorFromString("registerLibrary:withVersion:")) {
      FirebaseApp.perform(
        NSSelectorFromString("registerLibrary:withVersion:"),
        with: instance.firebaseLibraryName(),
        with: instance.firebaseLibraryVersion()
      )
    }
  }

  public func didReinitializeFirebaseCore(_ completion: @escaping () -> Void) {
    completion()
  }

  public func pluginConstants(for firebaseApp: FirebaseApp) -> [AnyHashable: Any] {
    let firebaseRemoteConfig = RemoteConfig.remoteConfig(app: firebaseApp)
    let configProperties = configProperties(for: firebaseRemoteConfig)
    var configValues: [String: Any] = configProperties
    configValues["parameters"] = getAllParameters(for: firebaseRemoteConfig)

    return configValues
  }

  func fetch(appName: String, completion: @escaping (Result<Void, any Error>) -> Void) {
    getRemoteConfig(from: appName).fetch { status, error in
      if let error {
        completion(.failure(self.createFlutterError(error)))
      } else {
        completion(Result.success(()))
      }
    }
  }

  func fetchAndActivate(appName: String, completion: @escaping (Result<Bool, any Error>) -> Void) {
    getRemoteConfig(from: appName).fetchAndActivate { status, error in
      if let error {
        completion(.failure(self.createFlutterError(error)))
      } else {
        completion(Result.success(status == .successFetchedFromRemote))
      }
    }
  }

  func activate(appName: String, completion: @escaping (Result<Bool, any Error>) -> Void) {
    getRemoteConfig(from: appName).activate { status, error in
      if let error {
        completion(.failure(self.createFlutterError(error)))
      } else {
        completion(Result.success(status))
      }
    }
  }

  func setConfigSettings(appName: String, settings: RemoteConfigPigeonSettings,
                         completion: @escaping (Result<Void, any Error>) -> Void) {
    let fetchTimeout = settings.fetchTimeoutSeconds
    let minFetchInterval = settings.minimumFetchIntervalSeconds
    let configSettings = RemoteConfigSettings()
    configSettings.fetchTimeout = Double(fetchTimeout)
    configSettings.minimumFetchInterval = Double(minFetchInterval)
    getRemoteConfig(from: appName).configSettings = configSettings
    completion(.success(()))
  }

  func setDefaults(appName: String, defaultParameters: [String: Any?],
                   completion: @escaping (Result<Void, any Error>) -> Void) {
    var filtered: [String: NSObject] = [:]

    for (key, value) in defaultParameters {
      if let nonNil = value, let obj = nonNil as? NSObject {
        filtered[key] = obj
      }
    }

    getRemoteConfig(from: appName).setDefaults(filtered)
    completion(.success(()))
  }

  func ensureInitialized(appName: String, completion: @escaping (Result<Void, any Error>) -> Void) {
    getRemoteConfig(from: appName).ensureInitialized { error in
      if let error {
        completion(.failure(self.createFlutterError(error)))
      } else {
        completion(.success(()))
      }
    }
  }

  func setCustomSignals(appName: String, customSignals: [String: Any?],
                        completion: @escaping (Result<Void, any Error>) -> Void) {
    let signalValues = convertToCustomSignalValues(customSignals)
    Task {
      do {
        try await getRemoteConfig(from: appName).setCustomSignals(signalValues)
        completion(.success(()))
      } catch {
        completion(.failure(createFlutterError(error)))
      }
    }
  }

  func getAll(appName: String, completion: @escaping (Result<[String: Any?], any Error>) -> Void) {
    let remoteConfig = getRemoteConfig(from: appName)
    let allKeys = Set(remoteConfig.allKeys(from: .static))
      .union(remoteConfig.allKeys(from: .default))
      .union(remoteConfig.allKeys(from: .remote))

    var parameters: [String: Any] = [:]
    for key in allKeys {
      let value = remoteConfig.configValue(forKey: key)
      parameters[key] = [
        "value": FlutterStandardTypedData(bytes: value.dataValue),
        "source": mapSource(value.source),
      ]
    }
    completion(.success(parameters))
  }

  func getProperties(appName: String,
                     completion: @escaping (Result<[String: Any], any Error>) -> Void) {
    let config = getRemoteConfig(from: appName)
    completion(.success(configProperties(for: config)))
  }

  public func firebaseLibraryName() -> String {
    "flutter-fire-rc"
  }

  public func firebaseLibraryVersion() -> String {
    versionNumber
  }

  public func flutterChannelName() -> String {
    kFirebaseRemoteConfigChannelName
  }

  public func onListen(withArguments arguments: Any?,
                       eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    guard let args = arguments as? [String: Any], let appName = args["appName"] as? String else {
      return nil
    }
    let remoteConfig = getRemoteConfig(from: appName)
    listenersMap[appName] = remoteConfig.addOnConfigUpdateListener { update, error in
      if let error {
        print("Remote Config update error: \(error.localizedDescription)")
        return
      }
      if let update {
        events(Array(update.updatedKeys))
      }
    }
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    guard let args = arguments as? [String: Any], let appName = args["appName"] as? String else {
      return nil
    }
    listenersMap[appName]?.remove()
    listenersMap.removeValue(forKey: appName)
    return nil
  }

  private func getRemoteConfig(from appName: String) -> RemoteConfig {
    let app = FLTFirebasePlugin.firebaseAppNamed(appName)
    return RemoteConfig.remoteConfig(app: app!)
  }

  private func getAllParameters(for remoteConfig: RemoteConfig) -> [String: Any] {
    var keySet = Set<String>()
    keySet.formUnion(remoteConfig.allKeys(from: .static))
    keySet.formUnion(remoteConfig.allKeys(from: .default))
    keySet.formUnion(remoteConfig.allKeys(from: .remote))

    var parameters: [String: Any] = [:]
    for key in keySet {
      parameters[key] = createRemoteConfigValueDict(remoteConfig.configValue(forKey: key))
    }

    return parameters
  }

  private func createRemoteConfigValueDict(_ remoteConfigValue: RemoteConfigValue)
    -> [String: Any] {
    [
      "value": FlutterStandardTypedData(bytes: remoteConfigValue.dataValue),
      "source": mapSource(remoteConfigValue.source),
    ]
  }

  private func mapSource(_ source: RemoteConfigSource) -> String {
    switch source {
    case .static: return "static"
    case .default: return "default"
    case .remote: return "remote"
    @unknown default: return "static"
    }
  }

  private func mapFetchStatus(_ status: RemoteConfigFetchStatus) -> String {
    switch status {
    case .success: return "success"
    case .failure: return "failure"
    case .throttled: return "throttled"
    case .noFetchYet: return "noFetchYet"
    @unknown default: return "failure"
    }
  }

  private func configProperties(for config: RemoteConfig) -> [String: Any] {
    [
      "fetchTimeout": Int(config.configSettings.fetchTimeout),
      "minimumFetchInterval": Int(config.configSettings.minimumFetchInterval),
      "lastFetchTime": Int(config.lastFetchTime?.timeIntervalSince1970 ?? 0 * 1000),
      "lastFetchStatus": mapFetchStatus(config.lastFetchStatus),
    ]
  }

  private func createFlutterError(_ error: Error) -> FlutterError {
    let nsError = error as NSError
    return FlutterError(
      code: "firebase_remote_config",
      message: nsError.localizedDescription,
      details: nsError.userInfo["details"]
    )
  }

  private func convertToCustomSignalValues(_ raw: [String: Any?]) -> [String: CustomSignalValue?] {
    raw.mapValues { value in
      guard let unwrapped = value else {
        return nil
      }

      switch unwrapped {
      case let string as String:
        return .string(string)
      case let int as Int:
        return .integer(int)
      case let double as Double:
        return .double(double)
      default:
        return nil
      }
    }
  }
}
